extensions/ipynb/src/notebookAttachmentCleaner.ts

```diff
@@ -0,0 +1,344 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { ATTACHMENT_CLEANUP_COMMANDID, JUPYTER_NOTEBOOK_MARKDOWN_SELECTOR } from './constants';
+import { DebounceTrigger, deepClone, objectEquals } from './helper';
+
+interface AttachmentCleanRequest {
+	notebook: vscode.NotebookDocument;
+	document: vscode.TextDocument;
+	cell: vscode.NotebookCell;
+}
+
+interface IAttachmentData {
+	[key: string /** mimetype */]: string;/** b64-encoded */
+}
+
+interface IAttachmentDiagnostic {
+	name: string;
+	ranges: vscode.Range[];
+}
+
+export enum DiagnosticCode {
+	missing_attachment = 'notebook.missing-attachment'
+}
+
+export class AttachmentCleaner implements vscode.CodeActionProvider {
+	private _attachmentCache:
+		Map<string /** uri */, Map<string /** cell fragment*/, Map<string /** attachment filename */, IAttachmentData>>> = new Map();
+
+	private _disposables: vscode.Disposable[];
+	private _imageDiagnosticCollection: vscode.DiagnosticCollection;
+	constructor() {
+		this._disposables = [];
+		const debounceTrigger = new DebounceTrigger<AttachmentCleanRequest>({
+			callback: (change: AttachmentCleanRequest) => {
+				this.cleanNotebookAttachments(change);
+			},
+			delay: 500
+		});
+
+		this._imageDiagnosticCollection = vscode.languages.createDiagnosticCollection('Notebook Image Attachment');
+		this._disposables.push(this._imageDiagnosticCollection);
+
+		this._disposables.push(vscode.commands.registerCommand(ATTACHMENT_CLEANUP_COMMANDID, async (document: vscode.Uri, range: vscode.Range) => {
+			const workspaceEdit = new vscode.WorkspaceEdit();
+			workspaceEdit.delete(document, range);
+			await vscode.workspace.applyEdit(workspaceEdit);
+		}));
+
+		this._disposables.push(vscode.languages.registerCodeActionsProvider(JUPYTER_NOTEBOOK_MARKDOWN_SELECTOR, this));
```

Comment:
When registering a code action provider, also pass in the third metadata which tells VS Code the kind of code actions you provide (in this case it would be `vscode.CodeActionKind.QuickFix`)

Reply:
:thumbsup: